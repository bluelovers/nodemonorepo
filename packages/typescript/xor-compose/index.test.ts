import compose, {Fn} from './index'

namespace fn {
  export const charCodeArray = (str: string) =>
    Array.from(str).map(x => x.charCodeAt(0))

  export const a = (str: string) => charCodeArray(str).reduce((a, b) => a | b)
  export const b = (str: string) => charCodeArray(str).reduce((a, b) => a & b)
  export const c = (str: string) => charCodeArray(str).reduce((a, b) => a ^ b)
  export const d = (str: string) => charCodeArray(str).reduce((a, b) => a + b)

  export const param = 'abcdefABCDEF'

  export const combinations: [string, Fn<string>][] = [
    ['a', compose(a)],
    ['b', compose(b)],
    ['c', compose(c)],
    ['d', compose(d)],
    ['a^b', compose(a, b)],
    ['b^c', compose(b, c)],
    ['c^d', compose(c, d)],
    ['d^a', compose(d, a)],
    ['a^c', compose(a, c)],
    ['b^d', compose(b, d)],
    ['a^b^c', compose(a, b, c)],
    ['a^b^d', compose(a, b, d)],
    ['a^c^d', compose(a, c, d)],
    ['b^c^d', compose(b, c, d)],
    ['a^b^c^d', compose(a, b, c, d)]
  ]
}

it('single composition should act like its sole provided function', () => {
  const call = (name: string) => {
    const res = fn.combinations.find(([x]) => x === name)
    return res ? res[1](fn.param) : undefined
  }

  expect({
    a: call('a'),
    b: call('b'),
    c: call('c'),
    d: call('d')
  }).toStrictEqual({
    a: fn.a(fn.param),
    b: fn.b(fn.param),
    c: fn.c(fn.param),
    d: fn.d(fn.param)
  })
})

it('matches snapshot', () => {
  const result = fn.combinations.map(
    ([name, func]) => ({name, result: func(fn.param)})
  )

  expect(result).toMatchSnapshot()
})
