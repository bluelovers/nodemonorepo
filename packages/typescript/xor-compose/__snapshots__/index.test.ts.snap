// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Array [
  Object {
    "name": "a",
    "result": 103,
  },
  Object {
    "name": "b",
    "result": 64,
  },
  Object {
    "name": "c",
    "result": 0,
  },
  Object {
    "name": "d",
    "result": 1002,
  },
  Object {
    "name": "a^b",
    "result": 39,
  },
  Object {
    "name": "b^c",
    "result": 64,
  },
  Object {
    "name": "c^d",
    "result": 1002,
  },
  Object {
    "name": "d^a",
    "result": 909,
  },
  Object {
    "name": "a^c",
    "result": 103,
  },
  Object {
    "name": "b^d",
    "result": 938,
  },
  Object {
    "name": "a^b^c",
    "result": 39,
  },
  Object {
    "name": "a^b^d",
    "result": 973,
  },
  Object {
    "name": "a^c^d",
    "result": 909,
  },
  Object {
    "name": "b^c^d",
    "result": 938,
  },
  Object {
    "name": "a^b^c^d",
    "result": 973,
  },
]
`;
